lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,505 @@
+  }
+
+  /**
+   * Signs in the user to Github using the Firebase API.
+   * @return {!Promise<object>} The logged in user.
+   */
+  signIn() {
+    return firebase.auth().signInWithPopup(this.provider).then(result => {
+      this.accessToken = result.credential.accessToken;
+      // A limitation of FB auth is that it doesn't return an oauth token
+      // after a page refresh. We'll get a firebase token, but not an oauth token
+      // for GH. Since GH's tokens never expire, stash the access token in IDB.
+      idb.set('accessToken', this.accessToken); // Note: async.
+      this.user = result.user;
+      return this.accessToken;
+    });
+  }
+
+  /**
+   * Signs the user out.
+   * @return {!Promise}
+   */
+  signOut() {
+    return firebase.auth().signOut().then(() => {
+      this.accessToken = null;
+      idb.delete('accessToken'); // Note: async.
+    });
+  }
+}
+
+class GithubAPI {
+  constructor() {
+    // this.CLIENT_ID = '48e4c3145c4978268ecb';
+    this.auth = new FirebaseAuth();
+  }
+
+  static get LH_JSON_EXT() {
+    return '.lighthouse.report.json';
+  }
+
+  authorize() {
+    return this.auth.accessToken ? Promise.resolve(this.auth.accessToken) :
+                                   this.auth.signIn();
+  }
+
+  /**
+   * Creates a gist under the users account.
+   * @param {!string} jsonFile The gist file body.
+   * @return {!Promise<string>} id of the created gist.
+   */
+  createGist(jsonFile) {
+    logger.log('Saving JSON to Github gist...', false);
+
+    return this.authorize()
+      .then(accessToken => {
+        // Stringify twice so quotes are escaped for POST request to succeed.
+        const filename = AssetSaver.getFilenamePrefix({url: jsonFile.url});
+        const content = JSON.stringify(JSON.stringify(jsonFile));
+        const body = `{
+          "description": "Lighthouse json report",
+          "public": false,
+          "files": {
+            "${filename}${GithubAPI.LH_JSON_EXT}": {
+              "content": ${content}
+            }
+          }
+        }`;
+
+        const request = new Request('https://api.github.com/gists', {
+          method: 'POST',
+          headers: new Headers({Authorization: `token ${accessToken}`}),
+          body
+        });
+        return fetch(request);
+      })
+      .then(resp => resp.json())
+      .then(json => {
+        logger.hide();
+        return json.id;
+      });
+  }
+
+  /**
+   * Fetches the body content of a gist.
+   * @param {!string} id The id of a gist.
+   * @return {!Promise<object>} json content of the gist.
+   */
+  getGistContent(id) {
+    return this.auth.ready.then(user => {
+      const headers = new Headers();
+
+      // If there's an authenticated user, include an Authorization header to
+      // have higher rate limits with the Github API. Otherwise, rely on ETags.
+      if (user) {
+        headers.set('Authorization', `token ${this.auth.accessToken}`);
+      }
+
+      return idb.get(id).then(gist => {
```

Comment:
maybe rename this to `cachedGist` or something just to differentiate from the gist you fetch below?

had to go back up to see that idb is local copy and it isn't set in this cache until farther down